File: c++/academia/postsecondary/Stack.cpp
// Description: This program demonstrates the implementation of a stack data structure in C++.

#include <iostream>
using namespace std;

class Stack {
    private:
    int size;
    int *stack;
    int top;

    public:
    Stack(int s) {
        size = s;
        stack = new int[size];
        top = -1;
    }

    ~Stack() {
        delete[] stack;
    }

    bool isEmpty() {
        return top == -1;
    }

    bool isFull() {
        return top == size - 1;
    }

    void push(int n) {
        if (!isFull()) {
            stack[++top] = n;
        } else {
            cout << "Stack is full!" << endl;
        }
    }

    int pop() {
        if (!isEmpty()) {
            return stack[top--];
        } else {
            cout << "Stack is empty!" << endl;
            return -1; // Returning -1 as an error value
        }
    }

    void display() {
        if (isEmpty()) {
            cout << "Stack is empty!" << endl;
            return;
        }

        for (int i = top; i >= 0; i--) {
            cout << stack[i] << " ";
        }
        cout << endl;
    }

    int getTopElement() {
        if (!isEmpty()) {
            return stack[top];
        } else {
            cout << "Stack is empty!" << endl;
            return -1; // Returning -1 as an error value
        }
    }
};
        
int main() {
    Stack stack(3);

    stack.push(10);
    stack.push(20);
    stack.push(30);

    stack.display(); // 30 20 10

    cout << "Top element: " << stack.getTopElement() << endl;

    return 0;
}

File: java/academia/uni-postsecondary/datastructures/Stack/Stack.java
package datastructures.Stack;
// Define a class called Stack that implements a stack data structure.

public class Stack<E> {
    private Node<E> top;
    private int size;

    public Stack() {
        this.top = null;
        this.size = 0;
    }

    public void push(E val) {
        Node<E> add = new Node<>(val);
        if(!empty()) {
            add.next = this.top;
        }
        this.top = add;      
        size++;
    }

    public E pop() {
        if (empty()) {
            return null;
        } else {
            E deletedVal = this.top.value;
            this.top = this.top.next; 
            size--;
            return deletedVal;
        }
    }

    public E peek() {
        if (empty()) {
            return null;
        }
        return this.top.value;
    }

    public int size() {
        return this.size;
    }

    public boolean empty() {
        return this.size == 0;
    }

    private class Node<E> {
        E value;
        Node<E> next;

        Node(E value) {
            this.value = value;
            this.next = null;
        }
    }
}